marlowe-playground-client/src/Simulation.purs

```diff
@@ -1,112 +1,506 @@
+  currContract <- editorGetValue
+  let
+    newContract = fromMaybe mempty $ oldContract <|> currContract
+  editorSetValue newContract
+  resetContract
+
+handleAction _ Undo = do
+  modifying _marloweState tailIfNotEmpty
+  mCurrContract <- use _currentContract
+  case mCurrContract of
+    Just currContract -> editorSetValue (show $ genericPretty currContract)
+    Nothing -> pure unit
+  where
+  tailIfNotEmpty ms =
+    let
+      tail = NEL.tail ms
+    in
+      case NEL.fromList tail of
+        Nothing -> ms
+        Just netail -> netail
+
+handleAction _ (SelectHole hole) = assign _selectedHole hole
+
+handleAction _ (InsertHole constructor firstHole holes) = do
+  mCurrContract <- editorGetValue
+  case mCurrContract of
+    Just currContract -> do
+      -- If we have a top level hole we don't want surround the value with brackets
+      -- so we parse the editor contents and if it is a hole we strip the parens
+      let
+        contractWithHole = case runParser hole currContract of
+          Right _ -> stripParens $ replaceInPositions constructor firstHole holes currContract
+          Left _ -> replaceInPositions constructor firstHole holes currContract
+
+        prettyContract = case runParser P.contract contractWithHole of
+          Right c -> show $ genericPretty c
+          Left _ -> contractWithHole
+      editorSetValue prettyContract
+    Nothing -> pure unit
+  where
+  stripParens s =
+    fromMaybe s
+      $ do
+          withoutPrefix <- stripPrefix (Pattern "(") $ trim s
+          withoutSuffix <- stripSuffix (Pattern ")") withoutPrefix
+          pure withoutSuffix
+
+handleAction _ (ChangeSimulationView view) = do
+  assign _bottomPanelView view
+  assign _showBottomPanel true
+  void $ query _editorSlot unit (Monaco.Resize unit)
+
+handleAction _ (ChangeHelpContext help) = do
+  assign _helpContext help
+  scrollHelpPanel
+
+handleAction settings CheckAuthStatus = do
+  checkAuthStatus settings
+
+handleAction settings (GistAction subEvent) = handleGistAction settings subEvent
+
+handleAction _ (ShowRightPanel val) = assign _showRightPanel val
+
+handleAction _ (ShowBottomPanel val) = do
+  assign _showBottomPanel val
+  void $ query _editorSlot unit (Monaco.Resize unit)
+
+handleAction _ (ShowErrorDetail val) = assign _showErrorDetail val
+
+handleAction _ SetBlocklyCode = do
+  source <- editorGetValue
+  case source of
+    Just source' -> H.raise (BlocklyCodeSet source')
+    Nothing -> pure unit
+
+handleAction _ AnalyseContract = do
+  currContract <- use _currentContract
+  currState <- use (_currentMarloweState <<< _state)
+  case currContract of
+    Nothing -> pure unit
+    Just contract -> do
+      checkContractForWarnings (encodeJSON contract) (encodeJSON currState)
+      assign _analysisState Loading
+  where
+  checkContractForWarnings contract state = do
+    let
+      msgString = unsafeStringify <<< encode $ CheckForWarnings contract state
+    H.raise (WebsocketMessage msgString)
```

Comment:
Would it be better if `WebsocketMessage` took a Foreign? 🤔

Reply:
I'm not sure, there's no restriction on the communication that suggests that the data has to be json, it just so happens that what's been used so far is JSON.